### ISSUE TITLE
Add support for BatchNormalization operation

### ISSUE THREAD

**Issue by JacksonZyy** [eth-sri/eran#40]
Hi,

I am very interested in your work and try to use the system to analyze an onnx network with BatchNormalization operations.
So I studied your code in ERAN and ELINA, and tried to implement the analysis on BatchNormalization (since BN layers function as affine layers during the inference phase, so I basically implemented BN analysis based on your code to handle Conv layers).
If I understand DeepPoly correctly, DeepPoly should function the same as directly executing the network when feeding with epsilon = 0. So if I implement support for BN layers correctly, I should see the same results from executing the network and executing DeepPoly with epsilon = 0. But the experiment result is not the case... So could you please advise me how can I test out whether I handle BN layers correctly?
Thanks for your reply in advance.

Best regards, 
Yuyi

**Comment by GgnDpSngh** [comment]
Hi Yuyi,

Thanks for your comment. Yes, your hypothesis is correct, you should get the same output scores up to at least the 5-th decimal when running ERAN with eps=0 as you would get when executing the network with PyTorch or Tensorflow. If you have your implementation of BatchNorm available anywhere, I can take a look.

Cheers,
Gagandeep Singh

**Comment by JacksonZyy** [comment]
Hi,

Thanks for your reply, I will consolidate my code and inform you once it is ready.
In the meantime, may I check with you that, with eps=0, not only the final output scores should be the same, the outputs for intermediate neurons should also be the same right? That is to say, the concrete lower bounds for neurons in intermediate layers should be the same as the concrete values when executing the network with PyTorch?

Thanks and regards,
Yuyi

**Comment by JacksonZyy** [comment]
Hi,

I've uploaded my codes to https://github.com/JacksonZyy/BN_supported_ERAN.
I add support of BN layers for deeppoly domain and onnx model.
I only uploaded those folders with the modified codes. 
More specifically, 
1. /tf_verify/onnx_translator.py is updated to parse BN layers
2. /tf_verify/optimizer.py is updated to add BN node to the execution list 
3. /tf_verify/deeppoly_nodes.py is updated with BN node and call out ELINA lib
4. /ELINA/fppoly/fppoly.c & fppoly.h are updated with handle_batchnormalization_layer function
5. /ELINA/python_interface/fppoly.py is updated to encapsulate BN C function to python function.
Thank you in advance!

Best regards,
Yuyi

**Comment by JacksonZyy** [comment]
Hi,

I successfully found out the reason by myself.
The input shape to the pytorch is NCHW, but the input shape should be changed to NHWC when feed into deeppoly.
Now deeppoly functions the same as executing the network with PyTorch. So my support for BN layers should be right.

Best regards,
Yuyi

**Comment by GgnDpSngh** [comment]
Hi Yuyi, 

Great, this is usually a source of problem, I am glad it works now. Please send us a pull request if your code can be integrated into ERAN.

Cheers,
Gagandeep Singh
